(RK-216) Make tarball module and environment sources require wrapper directories; be compatible with `pdk build`

reidmv opened the pull request:
This commit modifies the tarball module and environment types to expect wrapper directories.

Previously, a tarball was expected to contain content with no wrapper. E.g.

```
shell$ tar -tzf archive-1.2.3.tar.gz
./Puppetfile
./metadata.json
…
```

This is not typically how people create tarballs, and it is not how `pdk build` creates tarballs. Rather, it is typical for tarballs to have a "wrapper" directory, which helps avoid clutter when unpacking them. E.g.

```
shell$ tar -tzf archive-1.2.3.tar.gz
archive-1.2.3/Puppetfile
archive-1.2.3/metadata.json
archive-1.2.3/…
```

This commit updates the R10K tarball facilities to always expect a wrapper directory, and handle it appropriately.

Most notably, this change means that content downloaded directly from the forge, or created with `pdk build`, is automatically compatible with R10K's tarball module type.

This commit is not backwards compatible with previous versions of the tarball type(s). Open for discussion as to whether or not this is acceptable, given feature maturity (experimental?) and likely/possible adoption personas (innovators/experimenters vs. mainstream).

This behavior would constitute a finished implementation of RK-216.

reidmv committed:
Update tarball tests

reidmv committed:
Add a wrapper file ignore list to R10K::Tarball

Some sources will wrap all content in a wrapper directory (expected),
but include a header file in the tarball root (bad). Namely, Github will
include a file named pax_global_header in the provided tarball.

This is still a valid wrapper dir, but we have to ignore the header
file.

reidmv committed:
Add remove_wrapper_dir support to R10K::Tarball

This capability allows R10K::Tarball to remove a wrapping directory from
a tarball source with unpacking it, and when checking to see if an
unpacked directory is in sync.

This is enabling work for consuming module tarballs directly from `pdk
build` or the Forge, which typically places module content inside a
wrapper directory of the form <author>-<name>-<version>.

reidmv committed:
Make tarball modules, envs expect a wrapper dir

Modules packaged on the Forge and by `pdk build` package up module
content into a wrapper directory with the form
<author>-<name>-<version>. In order for R10K to be able to consume this
content directly as packaged, module tarball sources should expect there
to be a wrapper directory, and remove it when unpacking modules from
tarballs.

Make environments have the same expectation, for consistency.

reidmv committed:
Update documentation for Tarball types

reidmv committed:
Update CHANGELOG for Tarball changes

reidmv committed:
Replace .delete_prefix with .sub for 2.4 compat

This makes it harder to tell what's going on, but it's compatible with
Ruby 2.4

tvpartytonight commented on lib/r10k/tarball.rb:
```diff
@@ -70,28 +87,54 @@ def cache_basename
-    #
+    # 
     # @param target_dir [String] Where to unpack the tarball
+    # @param remove_wrapper_dir [Boolean] Whether or not to remove the wrapper
```
does this need to be added to the parameter list for `unpack`?

reidmv replied:
Ah; nope, that's an rdoc `@param` leftover from a previous implementation. `remove_wrapper_dir` is now an attr reader method. I will remove the `@param` note.

reidmv committed:
Code review: remove erroneous rdoc param ref

Remove erroneous rdoc param ref found during code review

reidmv committed:
Fix possible absoulte path bug

The R10K::Tarball#unpack method previously assumed that its input was an
absolute path, but did not validate that. If a caller gave a relative
path, this could have caused problems in the algorithm the method used
to remove the wrapper dir. This commit fixes that issue by calling
File#expand_path on the method's input.

Additionally, this commit adds a comment explaining why (and how) we're
removing the wrapper dir the way we are.
